### PR TITLE
Remove buzz

### LINKED ISSUE #951: Replace Buzz with PSR http-client
Refs https://github.com/sonata-project/SonataMediaBundle/issues/910

Today we are using both bundles for http connections. IMO we should drop Buzz and Guzzle and use the PSR http-client.

### PR DESCRIPTION
**Pull request by wbloszyk**
## Subject

I am targeting this branch, because this change provide BC-break.

Closes #951

## Changelog

```markdown
### Added 
- Added support for `nyholm/psr7` ^1.4
- Added `sonata.media.http.base_client` service for `Symfony\Component\HttpClient\Psr18Client`
- Added `sonata.media.http.base_message_factory` service for `Nyholm\Psr7\Factory\Psr17Factory`

### Changed
- Changed `__construct()` argument orders in `BaseVideoProvider`, `DailyMotionProvider`, `VimeoProvider` and `YouTubeProvider`

### Removed
- Remove support for `kriswallsmith/buzz`

### CODE REVIEW

File: UPGRADE-4.0.md [hunk 1]
```diff
@@ -7,7 +7,16 @@ UPGRADE FROM 3.x to 4.0
 
 ### Dependencies
 
- - Drop support for `sonata-project/datagrid-bundle` < 3.0.
+- Drop support for `kriswallsmith/buzz`
+
+  If you are using media provider(s) service with Guzzle, you MUST create a custom service based on the Buzz client and add it to configuration:
```
Comment (phansys) [inline UPGRADE-4.0.md:12]:
```suggestion
  If you are using media provider services with Guzzle, you MUST create a custom service based on the Buzz client and add it to configuration:
```

File: UPGRADE-4.0.md [hunk 2]
```diff
@@ -7,7 +7,16 @@ UPGRADE FROM 3.x to 4.0
...
+
+       sontata_media:
+           http:
+               client: 'your_custom.buzz_client'           # Psr\Http\Client\ClientInterface
```
Comment (phansys) [inline UPGRADE-4.0.md:16]:
```suggestion
               client: 'your_custom.buzz_client'  # Psr\Http\Client\ClientInterface
```

File: docs/reference/advanced_configuration.rst [hunk 3]
```diff
@@ -177,17 +177,6 @@ Full configuration options:
...
-            client:          'symfony_http_client'       # You need symfony/http-client for this
-            message_factory: 'nyholm.psr7.psr17_factory' # You need nyholm/psr7 for this
-
-        services:
-            symfony_http_client:
-                class: Symfony\Component\HttpClient\Psr18Client
-
-            nyholm.psr7.psr17_factory:
-                class: Nyholm\Psr7\Factory\Psr17Factory
+            client:          'sonata.media.http.base_client'          # You need symfony/http-client for this
```
Comment (phansys) [inline docs/reference/advanced_configuration.rst:181]:
```suggestion
            client:          'sonata.media.http.base_client'  # You need symfony/http-client for this
```
We can unalign the values in other PR.

File: src/Provider/BaseVideoProvider.php [hunk 4]
```diff
@@ -189,15 +159,12 @@ public function postRemove(MediaInterface $media): void
...
-    /*
+    /**
      * Get provider reference url.
      *
-     * @param MediaInterface $media
-     *
      * @return string
      */
-    // abstract public function getReferenceUrl(MediaInterface $media);
+    abstract public function getReferenceUrl(MediaInterface $media);
```
Comment (phansys) [inline src/Provider/BaseVideoProvider.php:167]:
```suggestion
    abstract public function getReferenceUrl(MediaInterface $media): string;
```

**Review by phansys — CHANGES_REQUESTED** [review]

File: src/Provider/BaseVideoProvider.php [hunk 5]
```diff
@@ -189,15 +159,12 @@ public function postRemove(MediaInterface $media): void
...
     }
 
-    // NEXT_MAJOR: Uncomment this method
-    /*
+    /**
      * Get provider reference url.
      *
-     * @param MediaInterface $media
-     *
      * @return string
```
Comment (phansys) [inline src/Provider/BaseVideoProvider.php:165]: This docblock can be safely removed.

Comment (wbloszyk) [reply src/Provider/BaseVideoProvider.php]: Done by php-cs-fixer.

File: UPGRADE-4.0.md [hunk 6]
```diff
@@ -7,7 +7,16 @@ UPGRADE FROM 3.x to 4.0
...
+- Drop support for `kriswallsmith/buzz`
+
+  If you are using media provider services with Guzzle, you MUST create a custom service based on the Buzz client and add it to configuration:
+
+       sontata_media:
+           http:
+               client: 'your_custom.buzz_client' # Psr\Http\Client\ClientInterface
+               message_factory: 'your_custom.message_facory' # Psr\Http\Message\RequestFactoryInterface
+
+- Bump support for `sonata-project/datagrid-bundle` < 3.0.
```
Comment (phansys) [inline UPGRADE-4.0.md:19]: Seems weird to use the term "bump" in conjunction with the `<` operator.

**Review by phansys — APPROVED** [review]
I'm approving, but please, update the PR's description.

**Review by phansys — APPROVED** [review]
About the changelog, "Change default configuration from ... to ..." means "Moved configuration node"?
BTW, the "Deprecated" section is empty.

**Review by jordisala1991 — DISMISSED** [review]

**Comment by wbloszyk** [comment]
> About the changelog, "Change default configuration from ... to ..." means "Moved configuration node"?
> BTW, the "Deprecated" section is empty.



> About the changelog, "Change default configuration from ... to ..." means "Moved configuration node"?
> BTW, the "Deprecated" section is empty.

After read it again it make no sense for me. I update it.

**Comment by phansys** [comment]
IMO:
>Changed `_constructor()` argument orders

should be:
>Changed `__construct()` arguments order

**Comment by wbloszyk** [comment]
@phansys @sonata-project/contributors Can we merge it?

**Review by phansys — DISMISSED** [review]

**Comment by wbloszyk** [comment]
@sonata-project/contributors Rebased. Can we merge?

**Review by phansys — APPROVED** [review]

**Review by jordisala1991 — APPROVED** [review]

**Comment by jordisala1991** [comment]
Needs 3.x merged to master

**Comment by wbloszyk** [comment]
> Needs 3.x merged to master

After rebase tests are working. 

@sonata-project/contributors Can we mege it now?

**Review by phansys — DISMISSED** [review]

File: composer.json [hunk 7]
```diff
@@ -47,6 +47,7 @@
         "symfony/finder": "^4.4 || ^5.2",
         "symfony/form": "^4.4",
         "symfony/framework-bundle": "^4.4",
+        "symfony/http-client": "^4.4 || ^5.2",
```
Comment (jordisala1991) [inline composer.json:50]: this dependency is already on require dev

Comment (wbloszyk) [reply composer.json:50]: Removed from `require-dev`.

**Review by phansys — APPROVED** [review]

**Review by jordisala1991 — DISMISSED** [review]

File: UPGRADE-4.0.md [hunk 8]
```diff
@@ -7,7 +7,16 @@ UPGRADE FROM 3.x to 4.0
 
 ### Dependencies
 
- - Drop support for `sonata-project/datagrid-bundle` < 3.0.
+- Drop support for `kriswallsmith/buzz`
+
+  If you are using media provider services with Guzzle, you MUST create a custom service based on the Buzz client and add it to configuration:
```
Comment (franmomu) [inline UPGRADE-4.0.md:12]:
```suggestion
  If you are using media provider services with Buzz, you MUST create a custom service based on the Buzz client and add it to configuration:
```
?

File: UPGRADE-4.0.md [hunk 9]
```diff
@@ -7,7 +7,16 @@ UPGRADE FROM 3.x to 4.0
...
+
+       sontata_media:
```
Comment (franmomu) [inline UPGRADE-4.0.md:14]:
```suggestion
       sonata_media:
```

File: UPGRADE-4.0.md (hunk 8 above)
Comment (wbloszyk) [reply UPGRADE-4.0.md]: Done.

File: UPGRADE-4.0.md (hunk 9 above)
Comment (wbloszyk) [reply UPGRADE-4.0.md]: Done.

**Review by franmomu — APPROVED** [review]

**Review by jordisala1991 — APPROVED** [review]

**Pull request merged**